src/engine.ts

```diff
@@ -417,6 +460,27 @@ export function initializePolyfill(updateCallback: () => void) {
         ];
       };
 
+      const updateShadowHostState: (
+        node: ShadowRoot,
+        state: LayoutState
+      ) => LayoutState = (node, state) => {
```

Comment:
You could make this `createShadowHostUpdater` that takes the `ShadowRoot` and returns a `LayoutStateProvider`, and move it to the top level scope.

Reply:
It's mostly a preference: I tend to prefer that functions describe what they do with a verb when possible.

`updateShadowHostState` is different from `defaultStateProvider`: `defaultStateProvider` _is_ just a state provider _value_ (i.e. a function that takes a `LayoutState` and returns a new `LayoutState`) that we can use directly. Refactoring `updateShadowHostState` to follow the same pattern as `createStateUpdater` would give you:

```ts
function updateShadowHostState(node: ShadowRoot) {
  return (state: LayoutState) => {
    const rootQueryDescriptors: ContainerConditionEntry[] = [];
    for (const adoptedStyleSheet of node.adoptedStyleSheets) {
      if (adoptedStyleSheet) {
        for (const query of getDescriptorsForStyleSheet(adoptedStyleSheet)) {
          rootQueryDescriptors.push([
            new Reference(query),
            QueryContainerFlags.None,
          ]);
        }
      }
    }
    return {
      ...state,
      conditions: rootQueryDescriptors,
    };
  };
}
```

Now, the name `updateShadowHostState` doesn't make much sense anymore, because it's no longer describing what the function does, which is return a new `LayoutStateProvider` given a `ShadowRoot` element, hence the suggestion that it's renamed to `createShadowHostUpdater` (though I now believe `createShadowRootStateProvider` would be a better name).

Hope that helps!

Reply:
implemented, thank you for sending this impl.